core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java

```diff
@@ -64,122 +64,146 @@
     if (primitiveCodec != null) {
       LOG.trace("[{}] Found matching primitive codec {}", logPrefix, primitiveCodec);
-      return safeCast(primitiveCodec);
+      return uncheckedCast(primitiveCodec);
     }
     for (TypeCodec<?> userCodec : userCodecs) {
       if (userCodec.accepts(cqlType)) {
         LOG.trace("[{}] Found matching user codec {}", logPrefix, userCodec);
-        return safeCast(userCodec);
+        return uncheckedCast(userCodec);
       }
     }
-    return safeCast(getCachedCodec(cqlType, null));
+    return uncheckedCast(getCachedCodec(cqlType, null, false));
   }
 
   @Override
   public <T> TypeCodec<T> codecFor(T value) {
     Preconditions.checkNotNull(value);
     LOG.trace("[{}] Looking up codec for object {}", logPrefix, value);
 
-    for (TypeCodec<?> primitiveCodec : PRIMITIVE_CODECS) {
+    for (TypeCodec<?> primitiveCodec : primitiveCodecs) {
       if (primitiveCodec.accepts(value)) {
         LOG.trace("[{}] Found matching primitive codec {}", logPrefix, primitiveCodec);
-        return safeCast(primitiveCodec);
+        return uncheckedCast(primitiveCodec);
       }
     }
     for (TypeCodec<?> userCodec : userCodecs) {
       if (userCodec.accepts(value)) {
         LOG.trace("[{}] Found matching user codec {}", logPrefix, userCodec);
-        return safeCast(userCodec);
+        return uncheckedCast(userCodec);
       }
     }
 
     if (value instanceof TupleValue) {
-      return safeCast(codecFor(((TupleValue) value).getType(), TupleValue.class));
+      return uncheckedCast(codecFor(((TupleValue) value).getType(), TupleValue.class));
     } else if (value instanceof UdtValue) {
-      return safeCast(codecFor(((UdtValue) value).getType(), UdtValue.class));
+      return uncheckedCast(codecFor(((UdtValue) value).getType(), UdtValue.class));
     }
 
     GenericType<?> javaType = inspectType(value);
     LOG.trace("[{}] Continuing based on inferred type {}", logPrefix, javaType);
-    return safeCast(getCachedCodec(null, javaType));
+    return uncheckedCast(getCachedCodec(null, javaType, true));
+  }
+
+  @Override
+  public <T> TypeCodec<T> codecFor(GenericType<T> javaType) {
+    LOG.trace("[{}] Looking up codec for Java type {}", logPrefix, javaType);
+    for (TypeCodec<?> primitiveCodec : primitiveCodecs) {
+      if (primitiveCodec.accepts(javaType)) {
+        LOG.trace("[{}] Found matching primitive codec {}", logPrefix, primitiveCodec);
+        return uncheckedCast(primitiveCodec);
+      }
+    }
+    for (TypeCodec<?> userCodec : userCodecs) {
+      if (userCodec.accepts(javaType)) {
+        LOG.trace("[{}] Found matching user codec {}", logPrefix, userCodec);
+        return uncheckedCast(userCodec);
+      }
+    }
+    return uncheckedCast(getCachedCodec(null, javaType, false));
   }
 
   // Not exposed publicly, this is only used for the recursion from
   // createCovariantCodec(GenericType)
   private TypeCodec<?> covariantCodecFor(GenericType<?> javaType) {
```

Comment:
This has to be protected, subclasses may need to call this method directly if they override `createCovariantCodec`.